crates/ruff_linter/src/rules/pydoclint/rules/check_docstring.rs

```diff
@@ -9,11 +9,98 @@ use ruff_python_semantic::{Definition, MemberKind, SemanticModel};
 use ruff_text_size::{Ranged, TextRange};
 
 use crate::checkers::ast::Checker;
-use crate::docstrings::sections::{SectionContexts, SectionKind};
+use crate::docstrings::sections::{SectionContext, SectionContexts, SectionKind};
 use crate::docstrings::styles::SectionStyle;
 use crate::registry::Rule;
 use crate::rules::pydocstyle::settings::Convention;
 
+/// ## What it does
+/// Checks for function docstrings that do not have a returns section.
+///
+/// ## Why is this bad?
+/// Docstrings missing return sections are a sign of incomplete documentation
+/// or refactors.
+///
+/// ## Example
+/// ```python
+/// def calculate_speed(distance: float, time: float) -> float:
+///     """Calculate speed as distance divided by time.
+///
+///     Args:
+///         distance: Distance traveled.
+///         time: Time spent traveling.
+///     """
+///     return distance / time
+/// ```
+///
+/// Use instead:
+/// ```python
+/// def calculate_speed(distance: float, time: float) -> float:
+///     """Calculate speed as distance divided by time.
+///
+///     Args:
+///         distance: Distance traveled.
+///         time: Time spent traveling.
+///
+///     Returns:
+///         Speed as distance divided by time.
+///     """
+///     return distance / time
+/// ```
+#[violation]
+pub struct DocstringMissingReturns;
+
+impl Violation for DocstringMissingReturns {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("`return` is not documented in docstring")
+    }
+}
+
+/// ## What it does
+/// Checks for function docstrings that have a returns section without
+/// needing one.
+///
+/// ## Why is this bad?
+/// Functions without an explicit return should not have a returns section
+/// in their docstrings.
+///
+/// ## Example
+/// ```python
+/// def say_hello(n: int) -> None:
+///     """Says hello to the user.
+///
+///     Args:
+///         n: Number of times to say hello.
+///
+///     Returns:
+///         Doesn't return anything.
+///     """
+///     for _ in range(n):
+///         print("Hello!")
+/// ```
+///
+/// Use instead:
+/// ```python
+/// def say_hello(n: int) -> None:
+///     """Says hello to the user.
+///
+///     Args:
+///         n: Number of times to say hello.
+///     """
+///     for _ in range(n):
+///         print("Hello!")
+/// ```
+#[violation]
+pub struct DocstringExtraneousReturns;
+
+impl Violation for DocstringExtraneousReturns {
+    #[derive_message_formats]
+    fn message(&self) -> String {
+        format!("Docstring should not have a returns section")
```

Comment:
As a user, I think it wouldn't be entirely clear to me **why** the docstring doesn't need to contain a return statement. Could we add some explanation around *because the function contains no return statement*.